webp/image-decoder/src/main/cpp/image_decoder.cpp

```diff
@@ -0,0 +1,240 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <android/imagedecoder.h>
+#include <android/data_space.h>
+#include "image_decoder.h"
+#include "image_scale_descriptor.h"
+#include "android_debug.h"
+#include <cassert>
+#include <thread>
+
+// Default displaying time image for each frame, used when switching from one image file to
+// the next image file at run time:
+//  - frames inside the animated files would provide the display time, so kDEFAULT_DURATION
+//    is not needed
+//  - non-animated image files will need this.
+const uint64_t kDEFAULT_DURATION = (2 * 1000000000UL);  // 2 second
+
+/*
+ * ImageDecoder constructor:
+ */
+ImageDecoder::ImageDecoder(const char** files, uint32_t count,
+                           DecodeSurfaceDescriptor *frameBuf,
+                           AAssetManager* mgr)
+    : assetMgr(mgr), curImgFile(nullptr), imgAsset(nullptr),
+      decoder(nullptr),
+      headerInfo(nullptr), frameInfo(nullptr), frameDuration(0UL),
+      decodeState(DecodeState::IDLE),
+      decodedImageStride(0), rotationAngle(0) {
+
+    for (auto i = 0; i < count; i++) {
+      fileNames.push(files[i]);
+    }
+
+    bufInfo = *frameBuf;
+    if (count && frameBuf) {
+        switch (bufInfo.format) {
+            case SurfaceFormat::SURFACE_FORMAT_RGB_565:
+                bytePerPixel = 2;
+                break;
+            case SurfaceFormat::SURFACE_FORMAT_RGBA_8888:
+            case SurfaceFormat::SURFACE_FORMAT_RGBX_8888:
+                bytePerPixel = 4;
+                break;
+            default:
+                ASSERT(false, "Unsupported display buffer format(%d) to %s",
+                       bufInfo.format, __FUNCTION__);
+                return;
+        }
+    }
+
+    memset(&targetRect, 0, sizeof(targetRect));
+}
+
+/**
+ * Report decoder is currently decoding frames.
+ * User should call this function before sending command to decode the
+ * next frame.
+ */
+bool ImageDecoder::IsBusy(void) {
+    return (decodeState == DecodeState::BUSY ? true : false);
+}
+/*
+ * DecodeNextImage():  start decode the next image or frame.
+ */
+bool ImageDecoder::DecodeNextImage(void) {
+
+    std::unique_lock<std::mutex> decodeLock(decodeMutex);
+
+    if(IsBusy())
+        return false;
+
+    int status;
+    if(!curImgFile) {
+        curImgFile = fileNames.front();
+        fileNames.pop();
+        // Open the asset with the give name from the APK's assets folder.
+        imgAsset =
+                AAssetManager_open(assetMgr, curImgFile, AASSET_MODE_BUFFER);
+        ASSERT(imgAsset, "%s does not exist in %s", curImgFile,
```

Comment:
Using ASSERT in an app is a problem because the app will crash and not recover from common errors like missing files.

Reply:
thanks for the comment: the sample at beta stage is to teach developers, if API failed, assert and stop right there to save debugging time and saving developer's time, the reported bug will be clean (rather than "the sample failed").